File: sepath-diff.py
#!/usr/bin/env python2
from collections import defaultdict
from itertools import chain
import numpy as np
import re
import optparse
import sys
import json


def get_base(unit='bit'):
    if unit == 'bit':
        log = np.log2
    elif unit == 'nat':
        log = np.log 
    elif unit in ('digit', 'dit'):
        log = np.log10  
    else:
        raise ValueError('The "unit" "%s" not understood' % unit)
    return log

def shannon_entropy(freq, unit='bit'):
    """Calculates the Shannon Entropy (H) of a frequency.
    
    Arguments:
    
        - freq (``numpy.ndarray``) A ``Freq`` instance or ``numpy.ndarray`` with 
          frequency vectors along the last axis.
        - unit (``str``) The unit of the returned entropy one of 'bit', 'digit' 
          or 'nat'.
    """
    log = get_base(unit)
    shape = freq.shape # keep shape to return in right shape
    Hs = np.ndarray(freq.size / shape[-1]) # place to keep entropies
    # this returns an array of vectors or just a vector of frequencies
    freq = freq.reshape((-1, shape[-1])) 
    # this makes sure we have an array of vectors of frequencies
    freq = np.atleast_2d(freq)
    # get fancy indexing
    positives = freq != 0.
    for i, (freq, idx) in enumerate(zip(freq, positives)):
        freq = freq[idx] # keep only non-zero
        logs = log(freq) # logarithms of non-zero frequencies
        Hs[i] = -np.sum(freq * logs)
    Hs.reshape(shape[:-1])
    return Hs
 

def jensen_shannon_divergence(freq, weights =None, unit='bit'):
    """
    Calculates the Jensen-Shannon Divergence (Djs) of two or more frequencies.
    The weights are for the relative contribution of each frequency vector. 
    
    Arguments:
    
        - freq (``numpy.ndarray``) rank-2 array of frequencies along the last dimension
        - weights (``numpy.ndarray``) A rank-1 array with a weight for each  frequency vector.
        - unit (``str``) see: the function ``shannon_entropy``.
    """
    if weights is not None:
        if len(freq) != len(weights):
            raise ValueError('The number of frequencies and weights do not match.')
    if freq.ndim != 2:
        raise ValueError('At least two frequencies in a rank-2 array expected.')
    weighted_average = np.average(freq, axis=0, weights=weights)
    H_avg_freq = shannon_entropy(weighted_average, unit)
    H_freq = shannon_entropy(freq, unit)
    avg_H_freq = np.average(H_freq, weights=weights)
    JSD = H_avg_freq - avg_H_freq
    return JSD

def parse_sep(fh):
    seps_counts = defaultdict(lambda: defaultdict(dict))
    for line in fh:
        gene_id, seps0, seps1, count = line.rstrip().split("\t")
        sep = [[], []]
        for i, sep_str in enumerate((seps0, seps1)):
            if sep_str:
                for se_str in sep_str.split("-"):
                    e, part = map(int, se_str.split(":"))
                    sep[i].append((e, part))
            sep[i] = tuple(sep[i])
        seps_counts[gene_id][tuple(sep)] = count
    return {"assigned":seps_counts}

def load_samples(fns, group_ids, format):
    sample_ids = [fn.rsplit("/", 1)[-1].rsplit(".", 1)[0] for fn in fns]
    sep_ids = zip(group_ids, sample_ids)
    sample_seps = {}
    for sep_id, fn in zip(sep_ids, fns):
        fn = os.path.abspath(os.path.expanduser(fn))
        if format == "json":
            sep = json.load(open(fn))
        else:
            sep = parse_sep(open(fn))
        sample_seps[sep_id] = sep["assigned"]
    return sample_seps

def seps_to_arrs(sample_seps):
    sample_ids = tuple(sample_seps.keys())
    # union of gene_ids across samples
    gene_ids = tuple(sorted(set(chain.from_iterable(sample_sep.keys() for sample_sep in sample_seps.values()))))
    sep_arrs = {}
    for gene_id in gene_ids:
        sep_ids = tuple(sorted(set(chain.from_iterable(sample_sep[gene_id] for sample_sep in sample_seps.values()))))
        sep_arr = np.zeros((len(sample_ids), len(sep_ids)), dtype=long)
        for row_id, sample_id in enumerate(sample_ids):
            for col_id, sep_id in enumerate(sep_ids):
                # make sure not to modify the defaultdicts
                if (gene_id in sample_seps[sample_id]) and (sep_id in sample_seps[sample_id][gene_id]):
                    sep_arr[row_id, col_id] = sample_seps[sample_id][gene_id][sep_id]
        sep_arrs[gene_id] = (sep_ids, sep_arr)
    return (sample_ids, sep_arrs)

def arr_valid(min_count, min_perc):
    #
    pass_count = arr >= min_count
    pass_freq = (arr.transpose() / np.array(arr.sum(axis=1), dtype=float) > min_freq).transpose()
    #
    arr_pass = pass_count & pass_freq
    row_pass = arr_pass.any(axis=1)






sep_counts = arr.sum(axis=0)
sample_counts = arr.sum(axis=1)



# def sep_jsd(arr):
    




# fns = ["~/tmp/test1.sep", "~/tmp/test2.sep", "~/tmp/test3.sep"]
# group_ids = ["A", "B", "C"]
# format = "sep"

# sample_seps = load_samples(fns, group_ids, format)
# a = seps_to_arrs(sample_seps)


# a = seps_to_arrs(seps)
    

#     print gene_id, len(sep_ids)
                 


# [0].keys()




# a["assigned"].keys()


#     for (row_idx, sep_id) in enumerate(sep_ids):
#         for (col_idx, col_id) in enumerate(
#         sep_arr[]




# seps1 = a["assigned"]["ENSG00000246228.2"]
# seps2 = a["assigned"]["ENSG00000246228.2"]
# seps3 = a["assigned"]["ENSG00000246228.2"]
# gene_seps = [seps1, seps2, seps3]



    


# a = parse_sep(open("t"))



if __name__ == "__main__":
    optParser = optparse.OptionParser(

        usage = "%prog [options] sep_file(s)",
        
        description = \
        "This script takes sepath-count output and identifies " +
        "differentially spliced genes.",
        
        epilog = \
        "Written by Marcin Cieslik (mcieslik@med.umich.edu) " +
        "Michigan Center for Translational Pathology (c) 2014 "
    )

    optParser.add_option("-v", "--verbose", action="store_true", dest="verbose",
                         help="run-time messages printed to stderr")

    optParser.add_option("-g", "--groups", type="string", dest="groups", 
                         help="")

    optParser.add_option("-f", "--format", type="string", dest="format", 
                         default="sep", help="input file format 'sep' or 'json'")

    if len(sys.argv) == 1:
        optParser.print_help()
        sys.exit(1)

    (opts, args) = optParser.parse_args()
    
    if len(args) < 2:
        optParser.print_help()
        sys.exit(1)

    groups = opts.groups.split(",") if opts.groups else map(str, range(1, 1+len(args))) 

    if len(groups) != len(args):
        sys.stderr.write("error: The length of 'groups' does not match the " +
                         "number of 'sep' files.\n")
        sys.exit(1)
    
    seps = load_seps(args, groups, opts.format)


    

File: sepath_diff.r
#!/usr/bin/env Rscript
library(plyr)
library(stringr)
library(optparse)
library(data.table)
# library(parallel)
# library(doMC)
# registerDoMC(detectCores())
library(MultinomialCI)

merge.list =function (x, y, ...) {
    if (length(x) == 0) 
        return(y)
    if (length(y) == 0) 
        return(x)
    i = match(names(y), names(x))
    i = is.na(i)
    if (any(i)) 
        x[names(y)[which(i)]] = y[which(i)]
    x
}

shannonEntropy = function(freq, base=2) {
    tmp = log(freq, base)
    tmp[is.infinite(tmp)] = NA
    Hs = -rowSums(tmp*freq, na.rm=TRUE)
    return(Hs)
}

JSD = function(freq, weights=rep(1/nrow(freq), nrow(freq)), base=2) {
    wmean = matrix(colSums(sweep(freq, 1, weights, "*")), nrow=1)
    H_avg_freq = shannonEntropy(wmean, base)    
    H_freq = shannonEntropy(freq, base)
    avg_H_freq = weighted.mean(H_freq, weights)
    H_avg_freq - avg_H_freq
}

JSDg = function(freq, groups, weights=rep(1/nrow(freq), nrow(freq)), base=2) {
    ## this the first term of the JSD
    wmean = matrix(colSums(sweep(freq, 1, weights, "*")), nrow=1)
    H_avg_freq = shannonEntropy(wmean, base)
    ## now rather than calculateing the avg shannon entroypy
    ## of each of the freqs we calculate mean frequency for
    ## each group and for it the shannon entropy the group Hs
    ## are weighted by the group size to give the second term
    ## of the JSD
    avg_H_freq = 0
    for (group in unique(groups)) {
        group_sel = (groups == group)
        freq_g = freq[group_sel, , drop=FALSE]
        weights_g = weights[group_sel]
        weights_freq_g = weights_g / sum(weights_g)
        wmean_g = matrix(colSums(sweep(freq_g, 1, weights_freq_g, "*")), nrow=1)
        H_avg_freq_g = shannonEntropy(wmean_g, base)
        avg_H_freq = avg_H_freq + sum(weights_g) * H_avg_freq_g
    }
    H_avg_freq - avg_H_freq
}

sepDivergence = function(sep_count, groups, alpha, sepcut, rowcut, colcut) {
    ## For each sample determine which seps are present above 'sepcut' in at least 'colcut'
    ## samples. And which samples contain at least 'rowcut' 
    ## at a 'alpha' false positive rate
    sep_count = as.matrix(sep_count)
    tmp = alply(sep_count, 1, function(row) {
        if (sum(row) >= length(row)) {
            multinomialCI(row, alpha)[,1]
        } else {
            rep(0, length(row))
        }
    }, .parallel=TRUE)
    low = splat(rbind)(tmp)
    ## filter
    lowcut = (low > sepcut)
    col_pass = colSums(lowcut) > (colcut * nrow(low))
    row_pass = rowSums(lowcut) > (rowcut * ncol(low))
    sep_count_pass = sep_count[row_pass,col_pass,drop=FALSE]
    groups_pass = groups[row_pass]
    ## transform rows of counts into rows of frequencies
    gene_sample_coverage = rowSums(sep_count_pass)
    sep_freq = sweep(sep_count_pass, 1, gene_sample_coverage, "/")
    ## calculate groupwise JSD
    JSDg(sep_freq, groups_pass)
}

loadSamples = function(fns, groups) {
    sample_data = ldply(fns, function(fn) {
        df = read.delim(fn, header=TRUE, stringsAsFactors=FALSE)
        df$sample = str_split(basename(fn), "\\.")[[1]][1]
        return(df)
    })
    samples = sapply(str_split(basename(fns), "\\."), "[", 1)
    sample_groups = data.frame(group=groups, stringsAsFactors=FALSE)
    rownames(sample_groups) = samples
    return(list(sample_data, sample_groups))
}

sepCounts = function(df) {
    samples = unique(df$sample)
    sep_counts = dlply(df, c("gene_id"), function(gdf) {
        ggdf = dlply(gdf, c("sample"), function(sdf) {
            gd = data.frame(matrix(sdf$total, ncol=nrow(sdf)))
            colnames(gd) = paste(sdf$sep1, sdf$sep2, sep="--")
            rownames(gd) = sdf$sample[[1]]
            return(gd)
        })
        res = splat(rbind.fill)(ggdf)
        res[is.na(res)] = 0
        rownames(res) = unlist(lapply(ggdf, row.names))
        res[setdiff(samples, rownames(res)),] = 0
        res = res[samples,]
        return(res)
    }, .parallel=TRUE)
    return(sep_counts)
}           

sepDivergences = function(samples, alpha=0.05, sepcut=0, colcut=1/3, rowcut=1/10) {
    sep_counts = sepCounts(samples[[1]])
    groups = samples[[2]]
    llply(sep_counts, function(sep_count) {
        gene_groups = groups[rownames(sep_count),]
        sepDivergence(sep_count, gene_groups, alpha=alpha, sepcut=sepcut, rowcut=rowcut, colcut=colcut)
    })
}
    
option_list = list(
    
    make_option(c("--groups"), type="character",
                help="comma-delimited groups associated with the samples [default: A,B,C...]",
                metavar="groups"),

    make_option(c("--out"), type="character", default="stdout",
                help="output file name [default: %default]",
                metavar="out"),

    make_option(c("--alpha"), type="numeric", default=0.05,
                help="confidence level [default: %default]",
                metavar="alpha"),

    make_option(c("--sepcut"), type="numeric", default=0,
                help="sep cut-off [default: %default]",
                metavar="cutoff"),

    make_option(c("--colcut"), type="numeric", default=0,
                help="col cut-off [default: %default]",
                metavar="cutoff"),

    make_option(c("--rowcut"), type="numeric", default=0,
                help="row cut-off [default: %default]",
                metavar="cutoff"),

    
    make_option(c("--fract"), type="numeric", default=0.33,
                help="fract [default: %default]",
                metavar="fract")
    
    )

parser = OptionParser("%prog [options] sep_files",

    description=c(
        "This script takes two or more 'sep_files' belonging to two or more 'groups' and estimates",
        "for each gene differences in splice-exon path distributions, which is a proxy for alternative",
        "splicing. Specifically, it calculates the excess Jensen-Shannon divergence"),

    epilogue=c(
        "Written by Marcin Cieslik (mcieslik@med.umich.edu) ",
        "Michigan Center for Translational Pathology (c) 2014 "),

    option_list=option_list
    )

opt =  parse_args(parser, positional_arguments=TRUE)

if (length(opt$args) < 2) {
    write("sepath-diff.r requires at least two 'sep' files.", stderr())
    write("Incorrect number of required positional arguments.\n", stderr())
    print_help(parser)
    quit("no", 1)
}

if (is.null(opt$options$groups)) {
    groups = LETTERS[1:length(opt$args)]
    write(paste("warning: groups not specified assigning each sample to separate groups:",
                paste(groups, collapse=","), "\n"), stderr())
} else {
    groups = str_split(opt$options$groups, ",")[[1]]
}

if (length(opt$args) != length(groups)) {
    write("Number of 'groups' does not match the number of 'sep' files.\n", stderr())
    print_help(parser)
    quit("no", 1)
}

samples = loadSamples(opt$args, groups)
sepdiv = sepDivergences(samples)
df = data.frame(SEPDIV=as.matrix(unlist(sepdiv)))

if (opt$options$out == "stdout") {
    write.table(df, stdout(), sep="\t", quote=FALSE, col.names=FALSE)
} else {
    write.table(df, opt$options$out, sep="\t", quote=FALSE, col.names=FALSE)
}
